### PR TITLE
Fix project file name ordering with diamond dependencies

### PR DESCRIPTION
**Pull request by dsyme**
Fixes this problem https://github.com/Microsoft/visualfsharp/issues/1362 by propagating [this fix from FSharp.Compiler.Service](https://github.com/fsharp/FSharp.Compiler.Service/pull/594)

This is an important fix for a case in transitive dependencies that seems to have been missed.

### CODE REVIEW

**Comment by dsyme** [comment]
The whole of ``ProjectOptionsTests.fs`` from FSharp.Compiler.Service is included, even though only the last test is actually active.  The rest of the tests are for the ProjectCracker tool used by FSharp.Compiler.Service.

**Comment by dsyme** [comment]
WE have failing tests after this change:

```
InteractiveSession\Misc (LoadingFsx.fsx) -- failed
InteractiveSession\Misc (LoadingFsx.fsscript) -- failed

1) Failed : FSharp-Tests-Core+Attributes.attributes(FSI_FILE)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsiAnyCPU.exe' with args ' -r:System.Core.dll --nowarn:20 --define:INTERACTIVE --maxerrors:1 --abortonerror testlib.fsi testlib.fs test.fsx' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\attributes'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

2) Failed : FSharp-Tests-Core+FSI-reload.fsi-reload
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsiAnyCPU.exe' with args ' -r:System.Core.dll --nowarn:20 --define:INTERACTIVE --maxerrors:1 --abortonerror  --maxerrors:1 load2.fsx' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\fsi-reload'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

3) Failed : FSharp-Tests-Core+Load-Script.load-script
'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\load-script\out.stdout.txt' and 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\load-script\out.stdout.bsl' differ; ["diff between [D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\load-script\out.stdout.txt] and [D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\load-script\out.stdout.bsl]";
 "line 45"; " - Test 10=================================================";
 " + Hello"]
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

4) Failed : FSharp-Tests-Core+Members+Basics.Basics(FSI_FILE)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsiAnyCPU.exe' with args ' -r:System.Core.dll --nowarn:20 --define:INTERACTIVE --maxerrors:1 --abortonerror test.fsi test.fs' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\members\basics'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

5) Failed : FSharp-Tests-Core+Reflect.reflect(FSI_FILE)
Error running command 'D:\j\workspace\release_ci_pa---866fd2c3\tests\..\release\net40\bin\fsiAnyCPU.exe' with args ' -r:System.Core.dll --nowarn:20 --define:INTERACTIVE --maxerrors:1 --abortonerror test.fs test2.fs' in directory 'D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\core\reflect'. ERRORLEVEL 1 ERRORLEVEL 1
at NUnitConf.checkTestResult(Result`2 result) in D:\j\workspace\release_ci_pa---866fd2c3\tests\fsharp\nunitConf.fs:line 14

```

**Comment by dsyme** [comment]
The investigation of these failures is in the comments of https://github.com/fsharp/FSharp.Compiler.Service/pull/594.  We obviously need to fix them before we can consider this a good fix.

File: src/fsharp/CompileOps.fs [hunk 1]
```diff
@@ -5042,7 +5045,7 @@ module private ScriptPreprocessClosure =
     /// Used from fsi.fs and fsc.fs, for #load and command line
     let GetFullClosureOfScriptFiles(tcConfig:TcConfig,files:(string*range) list,codeContext,_useDefaultScriptingReferences:bool,lexResourceManager:Lexhelp.LexResourceManager) = 
         let mainFile = fst (List.head files)
-        let protoClosure = files |> List.map (fun (filename,m)->SourceFileOfFilename(filename,m,tcConfig.inputCodePage)) |> List.concat |> List.rev // Reverse to put them in the order they will be extracted later
+        let protoClosure = files |> List.map (fun (filename,m)->SourceFileOfFilename(filename,m,tcConfig.inputCodePage)) |> List.concat // Reverse to put them in the order they will be extracted later
```
Comment (alfonsogarciacaro) [inline src/fsharp/CompileOps.fs]: As `List.rev` is now gone, should the comment be deleted too?

**Comment by dsyme** [comment]
Now a different set of failure :(

```

LoadOrderOfExecution4.fs(1,9): error FS0039: The value or constructor 'y' is not defined
Compile Unexpectedly Failed: 256 
FAIL
+++ InteractiveSession\Misc (RelativeHashRResolution04_fsc) +++
Compiling: [fsc   --nologo aaa\bbb\RelativeHashRResolution04_1.fsx ]

aaa\bbb\RelativeHashRResolution04_1.fsx(8,14): error FS0039: The namespace or module 'RelativeHashRResolution04_2' is not defined
Compile Unexpectedly Failed: 256 
FAIL
+++ InteractiveSession\Misc (RelativeHashRResolution04_fscrelative) +++
Compiling: [fsc   --nologo ..\Misc\aaa\bbb\RelativeHashRResolution04_1.fsx ]

aaa\bbb\RelativeHashRResolution04_1.fsx(8,14): error FS0039: The namespace or module 'RelativeHashRResolution04_2' is not defined
Compile Unexpectedly Failed: 256 
FAIL
+++ InteractiveSession\Misc (RelativeHashRResolution04_fscsimple) +++
Compiling: [fsc   --nologo --simpleresolution --noframework -r:"D:\j\workspace\release_ci_pa---3f142ccc\tests\..\release\net40\bin\FSharp.Core.dll" aaa\bbb\RelativeHashRResolution04_1.fsx ]

aaa\bbb\RelativeHashRResolution04_1.fsx(8,14): error FS0039: The namespace or module 'RelativeHashRResolution04_2' is not defined
Compile Unexpectedly Failed: 256 
FAIL
+++ InteractiveSession\Misc (RelativeHashRResolution04_fscrelativesimple) +++
Compiling: [fsc   --nologo --simpleresolution --noframework -r:"D:\j\workspace\release_ci_pa---3f142ccc\tests\..\release\net40\bin\FSharp.Core.dll" ..\Misc\aaa\bbb\RelativeHashRResolution04_1.fsx ]

aaa\bbb\RelativeHashRResolution04_1.fsx(8,14): error FS0039: The namespace or module 'RelativeHashRResolution04_2' is not defined
Compile Unexpectedly Failed: 256 
FAIL


```

**Comment by dsyme** [comment]
@alfonsogarciacaro  Reworking the fix now

**Pull request merged**